### PR TITLE
chore: refactor build artifacts methods

### LINKED ISSUE #695: Implement local mode for metricbeat
At this moment, it's only supported for Fleet and the elastic-agent, not for metricbeat

### PR DESCRIPTION
**Pull request by mdelapenya**
Sorry for this massive PR, I started small but you know.. yak shaving

## What does this PR do?
This PR simplifies the generation of artifact names, adding unit tests and refactoring the code. More in detail:

- we are adding a helper method to generate build artifact names, removing duplicated code and isolation that functionality in one place, allowing to unit test it. It will support calculating the name for both local paths, GCP buckets and official repositories.
- we are moving the code around to always pass the artifact name, generated with above method, to avoid passing all tokens everywhere: artifact, arch, os, extension, etc.
- we are updating the artifact name for those artifacts downloaded from Elastic's official repository, because the names are different: they use a `-docker-image` token that comes right after the version, but in Beats build it uses a `.docker` token that comes right before the extension.
- on the same hand, we are considering that the artifact name could contain "-ubi8" as suffix, which only affects the artifacts to be downloaded.
- we are adding a new installer for Docker images, which is able to load the TAR file representing the image into the local docker engine at test time. It will use base TAR installer, and will be used in the standalone mode.
- we are consuming local docker images for Fleet's standalone mode, allowing running the standalone tests with local Beats repo (we include unit tests for it)
- we are consuming local docker images for Metricbeat test suite, loading the local images into the docker engine before metricbeat container is run. See #695 
- as a consequence of the above item, we are extracting the common logic to load a docker image to a helper method, so that metricbeat and the agent are able to load images when needed.
- we are also adding support for metricbeat to consume docker images from GCP snapshots, affecting Beats' PRs and branches. It always used the official artifacts, and now it's possible to consume a HASH, a pull request, or a CI snapshot, as we already do for Fleet. This change could have come in another PR, but is was so easy to add after this PR's refactors.
- we are extracting the download artifact to a helper method, so both metricbeat and the agent are able to download artifacts from GCP, local file system, or official repositories. This is a big change, in which we noticed that there is no need to pass the `stale` argument to the method calls: we set the version when creating the stale installer and that's all. We are checking that the version used is a PR at installer build time, 
- we removed an old log entry that was misleading
- we removed a wrong field in a log entry, which was causing logrus to complain.

## Why is it important?
Code health, avoiding redundant code and reusability.

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have run the Unit tests for the CLI and E2E, and they are passing locally
- [x] I have run the End-2-End tests for the suite I'm working on, and they are passing locally
- [ ] I have noticed new Go dependencies (run `make notice` in the proper directory)

## Author's Checklist

- [x] Run the tests for PRs, snapshots and releases
- [x] Run the tests for stale agents (Fleet's upgrade test) 

## How to test this PR locally
Let's use the existing goals:

```shell
make -C e2e fleet-fleet                   # Fleet test suite for fleet mode
make -C e2e fleet-fleet-ci-snapshot       # Fleet test suite for fleet mode for a SHA1
make -C e2e fleet-fleet-pr-ci-snapshot    # Fleet test suite for fleet mode for a PR
```

## Related issues

- Built on top of #688
- Closes #695

### CODE REVIEW

**Review by kuisathaverat — APPROVED** [review]

File: e2e/utils.go [hunk 1]
```diff
@@ -30,6 +30,21 @@ const charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
...
 
+// BuildArtifactName builds the artifact name from the different coordinates for the artifact
+func BuildArtifactName(artifact string, version string, OS string, arch string, extension string, isDocker bool) string {
+	dockerString := ""
+	if isDocker {
+		dockerString = ".docker"
+	}
+
+	artifactName := fmt.Sprintf("%s-%s-%s-%s%s.%s", artifact, version, OS, arch, dockerString, extension)
+	if extension == "deb" || extension == "rpm" {
```
Comment (cachedout) [inline e2e/utils.go:41]: Maybe lowercase the string before the comparison, just to be safe?

**Review by cachedout — APPROVED** [review]

File: e2e/utils.go (hunk 1 above)
Comment (mdelapenya) [reply e2e/utils.go]: Resolved in https://github.com/elastic/e2e-testing/pull/689/commits/4bee7c1b7e04e715b85379e0476dc633434d6bf9

**Review by cachedout — APPROVED** [review]

File: cli/docker/docker.go [hunk 2]
```diff
@@ -172,6 +175,33 @@ func RemoveContainer(containerName string) error {
 	return nil
 }
 
+// LoadImage loads a TAR file in the loall docker engine
```
Comment (mdelapenya) [inline cli/docker/docker.go:178]:
```suggestion
// LoadImage loads a TAR file in the local docker engine
```

File: e2e/_suites/fleet/services.go [hunk 3]
```diff
@@ -1,24 +1,19 @@
...
 	"github.com/elastic/e2e-testing/cli/docker"
-	"github.com/elastic/e2e-testing/cli/shell"
 	"github.com/elastic/e2e-testing/e2e"
 	log "github.com/sirupsen/logrus"
 )
 
-// to avoid downloading the same artifacts, we are adding this map to cache the URL of the downloaded binaries, using as key
-// the URL of the artifact. If another installer is trying to download the same URL, it will return the location of the
-// already downloaded artifact.
-var binariesCache = map[string]string{}
```
Comment (mdelapenya) [inline e2e/_suites/fleet/services.go:20]: Moved to utils

File: e2e/_suites/fleet/services.go [hunk 4]
```diff
@@ -166,84 +161,17 @@ func runElasticAgentCommand(profile string, image string, service string, proces
 // to be used will be defined by the local snapshot produced by the local build.
 // Else, if the environment variable BEATS_USE_CI_SNAPSHOTS is set, then the artifact
 // to be downloaded will be defined by the latest snapshot produced by the Beats CI.
-func downloadAgentBinary(artifact string, version string, OS string, arch string, extension string, stale bool) (string, string, error) {
```
Comment (mdelapenya) [inline e2e/_suites/fleet/services.go:169]: Moved (and abstracted) to utils

File: e2e/_suites/fleet/services.go [hunk 5]
```diff
@@ -166,84 +161,17 @@ func runElasticAgentCommand(profile string, image string, service string, proces
...
-		return handleDownload(downloadURL, bucketFileName)
-	}
-
-	downloadVersion := version
-	if !stale {
-		downloadVersion = checkElasticAgentVersion(version)
-	}
-
-	downloadURL, err = e2e.GetElasticArtifactURL(artifact, downloadVersion, OS, arch, extension)
+func downloadAgentBinary(artifactName string, artifact string, version string) (string, error) {
```
Comment (mdelapenya) [inline e2e/_suites/fleet/services.go:164]: I could have remove this method, but instead I decide to internally call the new method

File: e2e/_suites/fleet/services.go [hunk 6]
```diff
@@ -277,55 +209,12 @@ func GetElasticAgentInstaller(image string, installerType string, version string
 	return installer
 }
 
-// getGCPBucketCoordinates it calculates the bucket path in GCP
-func getGCPBucketCoordinates(fileName string, artifact string, version string, OS string, arch string, extension string, stale bool) (string, string, string, string) {
```
Comment (mdelapenya) [inline e2e/_suites/fleet/services.go:281]: Moved to utils

File: e2e/_suites/fleet/stand-alone.go [hunk 7]
```diff
@@ -62,6 +63,17 @@ func (sats *StandAloneTestSuite) contributeSteps(s *godog.ScenarioContext) {
...
 
+	useCISnapshots := shell.GetEnvBool("BEATS_USE_CI_SNAPSHOTS")
+	beatsLocalPath := shell.GetEnv("BEATS_LOCAL_PATH", "")
+	if useCISnapshots || beatsLocalPath != "" {
+		// load the docker images that were already:
+		// a. downloaded from the GCP bucket
+		// b. fetched from the local beats binaries
+		dockerInstaller := GetElasticAgentInstaller("docker", image, agentVersion)
+
+		dockerInstaller.PreInstallFn()
```
Comment (mdelapenya) [inline e2e/_suites/fleet/stand-alone.go:74]: Now it's possible to consume local or CI snapshots for the standalone mode 🚀

File: e2e/_suites/metricbeat/metricbeat_test.go [hunk 8]
```diff
@@ -313,6 +314,22 @@ func (mts *MetricbeatTestSuite) installedUsingConfiguration(configuration string
 
 // runMetricbeatService runs a metricbeat service entity for a service to monitor it
 func (mts *MetricbeatTestSuite) runMetricbeatService() error {
+	useCISnapshots := shell.GetEnvBool("BEATS_USE_CI_SNAPSHOTS")
+	beatsLocalPath := shell.GetEnv("BEATS_LOCAL_PATH", "")
+	if useCISnapshots || beatsLocalPath != "" {
```
Comment (mdelapenya) [inline e2e/_suites/metricbeat/metricbeat_test.go:319]: Now it's possible to load local or CI snapshots for metricbeat images!

File: e2e/utils.go [hunk 9]
```diff
@@ -53,7 +194,7 @@ func GetExponentialBackOff(elapsedTime time.Duration) *backoff.ExponentialBackOf
 // GetElasticArtifactVersion returns the current version:
 // 1. Elastic's artifact repository, building the JSON path query based
 // If the version is a PR, then it will return the version without checking the artifacts API
-// i.e. GetElasticArtifactVersion("8.0.0-SNAPSHOT")
+// i.e. GetElasticArtifactVersion("$VERSION")
```
Comment (mdelapenya) [inline e2e/utils.go:197]: This change will simplify maintenance in backports

File: e2e/utils.go [hunk 10]
```diff
@@ -127,14 +268,12 @@ func GetElasticArtifactVersion(version string) string {
...
-// 2. Elastic's artifact repository, building the JSON path query based
-// i.e. GetElasticArtifactURL("elastic-agent", "8.0.0-SNAPSHOT", "linux", "x86_64", "tar.gz")
-// i.e. GetElasticArtifactURL("elastic-agent", "8.0.0-SNAPSHOT", "x86_64", "rpm")
-// i.e. GetElasticArtifactURL("elastic-agent", "8.0.0-SNAPSHOT", "amd64", "deb")
-func GetElasticArtifactURL(artifact string, version string, operativeSystem string, arch string, extension string) (string, error) {
+// GetElasticArtifactURL returns the URL of a released artifact, which its full name is defined in the first argument,
+// from Elastic's artifact repository, building the JSON path query based on the full name
+// i.e. GetElasticArtifactURL("elastic-agent-$VERSION-amd64.deb", "elastic-agent", "$VERSION")
+// i.e. GetElasticArtifactURL("elastic-agent-$VERSION-x86_64.rpm", "elastic-agent","$VERSION")
+// i.e. GetElasticArtifactURL("elastic-agent-$VERSION-linux-amd64.tar.gz", "elastic-agent","$VERSION")
```
Comment (mdelapenya) [inline e2e/utils.go:275]: This change will simplify maintenance in backports

File: e2e/_suites/fleet/ingest-manager_test.go [hunk 11]
```diff
@@ -231,21 +233,6 @@ func (imts *IngestManagerTestSuite) processStateOnTheHost(process string, state
 	return checkProcessStateOnTheHost(containerName, process, state)
 }
 
-// checkElasticAgentVersion returns a fallback version (agentVersionBase) if the version set by the environment is empty
-func checkElasticAgentVersion(version string) string {
```
Comment (mdelapenya) [inline e2e/_suites/fleet/ingest-manager_test.go:235]: Moved (and abstracted) to utils

File: e2e/utils_test.go [hunk 12]
```diff
@@ -38,6 +40,176 @@ func init() {
...
+		artifactName = BuildArtifactName(artifact, version, fallbackVersion, OS, arch, "TAR.GZ", true)
+		assert.Equal(t, expectedFileName, artifactName)
+	})
+}
+
+func TestCheckPRVersion(t *testing.T) {
+	t.Run("Checking a version should return the version", func(t *testing.T) {
+		v := CheckPRVersion(testVersion, testVersion)
+
+		assert.Equal(t, "8.0.0-SNAPSHOT", v)
```
Comment (mdelapenya) [inline e2e/utils_test.go:202]:
```suggestion
		assert.Equal(t, testVersion, v)
```

**Comment by mdelapenya** [comment]
This is ready. I'm going to merge it right after the CI passes.

**Pull request merged**